swarm/storage/resource.go

```diff
@@ -0,0 +1,529 @@
+		ethapi:       ethapi,
+		resources:    make(map[string]*resource),
+		resourceLock: sync.Mutex{},
+		hashLock:     sync.Mutex{},
+		hasher:       hasher(),
+		privKey:      privKey,
+	}, nil
+}
+
+// Creates a standalone resource object
+//
+// Can be passed to SetResource if external root data lookups are used
+func NewResource(name string, startblock uint64, frequency uint64) (*resource, error) {
+	validname, err := idna.ToASCII(name)
+	if err != nil {
+		return nil, err
+	}
+
+	return &resource{
+		name:       validname,
+		ensname:    ens.EnsNode(validname),
+		startblock: startblock,
+		frequency:  frequency,
+	}, nil
+}
+
+// Creates a new root entry for a resource update identified by `name` with the specified `frequency`.
+//
+// The start block of the resource update will be the actual current block height of the connected network.
+func (self *ResourceHandler) NewResource(name string, frequency uint64) (*resource, error) {
+
+	// frequency 0 is invalid
+	if frequency == 0 {
+		return nil, fmt.Errorf("Frequency cannot be 0")
+	}
+
+	// make sure our ens identifier is idna safe
+	validname, err := idna.ToASCII(name)
+	if err != nil {
+		return nil, err
+	}
+	ensname := ens.EnsNode(validname)
+
+	// get our blockheight at this time
+	currentblock, err := self.getBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	// chunk with key equal to namehash points to data of first blockheight + update frequency
+	// from this we know from what blockheight we should look for updates, and how often
+	chunk := NewChunk(Key(ensname[:]), nil)
+	chunk.SData = make([]byte, INDEX_SIZE)
+
+	// resource update root chunks follow same convention as "normal" chunks
+	// with 8 bytes prefix specifying size
+	val := make([]byte, 8)
+	chunk.SData[0] = 16 // size, little-endian
+	binary.LittleEndian.PutUint64(val, currentblock)
+	copy(chunk.SData[8:16], val)
+	binary.LittleEndian.PutUint64(val, frequency)
+	copy(chunk.SData[16:], val)
+	self.Put(chunk)
+	log.Debug("new resource", "name", validname, "key", ensname, "startblock", currentblock, "frequency", frequency)
+
+	self.resourceLock.Lock()
+	defer self.resourceLock.Unlock()
+	self.resources[name] = &resource{
+		name:       validname,
+		ensname:    ensname,
+		startblock: currentblock,
+		frequency:  frequency,
+		updated:    time.Now(),
+	}
+	return self.resources[name], nil
+}
+
+// Set an externally defined resource object
+//
+// If the resource update root chunk is located externally (for example as a normal
+// chunk looked up by ENS) the data would be manually added with this method).
+//
+// Method will fail if resource is already registered in this session, unless
+// `allowOverwrite` is set
+func (self *ResourceHandler) SetResource(rsrc *resource, allowOverwrite bool) error {
+
+	if rsrc.name == "" {
+		return fmt.Errorf("Resource name cannot be empty")
+	}
+
+	utfname, err := idna.ToUnicode(rsrc.name)
+	if err != nil {
+		return fmt.Errorf("Invalid IDNA rsrc name '%s'", rsrc.name)
+	}
```

Comment:
Here we check the resource name again, let's refactor this into the function I mentioned in https://github.com/ethersphere/go-ethereum/pull/139/files#r156192771

Reply:
This check should be done in the constructor instead.